File: graphs & tries/1519.js
class Solution {
    dfs(adj, curr ,parent, result, labels, count){
        const myLabel = labels[curr]
        const countBeforeVisitingChild = count[myLabel.charCodeAt(0) - 'a'.charCodeAt(0)]
        count[myLabel.charCodeAt(0) - 'a'.charCodeAt(0)] += 1
        for(const v of adj[curr]){
            if(v === parent){
                continue
            }
            this.dfs(adj, v, curr, result, labels, count)
        }
        const countAfterVisitingChild = count[myLabel.charCodeAt(0) - 'a'.charCodeAt(0)]
        result[curr] = countAfterVisitingChild - countBeforeVisitingChild
    }
    countSubTress(n, edges, labels){
        const adj = Array.from({length: n}, () =>[])
        for(const [u, v] of edges){
            adj[u].push(v)
            adj[v].push(u)
        }
        const result = new Array(n).fill(0)
        const count = new Array(26).fill(0)
        this.dfs(adj, 0, -1, result, labels, count)
        return result
    }

}

File: graphs & tries/1519.py
# DFS
# Tc = O(V+E) = O(n + (n-1)) = O(n)
# Sc = O(N) [adjacency list]
from collections import defaultdict

class Solution:
    def dfs(self, adj, curr, parent, result, labels, count):
        my_label = labels[curr]
        count_before_visiting_curr_children = count[ord(my_label) - ord('a')]
        
        # Now will explore
        count[ord(my_label) - ord('a')] += 1
        
        for v in adj[curr]:
            if v == parent:
                continue
            self.dfs(adj, v, curr, result, labels, count)
        
        # Time to write after variable
        count_after_visiting_curr_children = count[ord(my_label) - ord('a')]
        result[curr] = count_after_visiting_curr_children - count_before_visiting_curr_children

    def countSubTrees(self, n, edges, labels):
        adj = defaultdict(list)
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        
        result = [0] * n
        count = [0] * 26
        self.dfs(adj, 0, -1, result, labels, count)
        return result
